Support for Alerts on Questions

senior opened the pull request:
This will allow users to be notified when a given question has results or if the question has met a given goal. The notification is similar to a pulse (email and slack).

Backend TODOs
-------------------------------
- [x] Creating alerts
- [x] Updating alerts
- [x] Emailing when rows are present
- [x] Emailing when data is above/below goal
- [x] Fix transient failure in pulse_channel_test
- [x] Add GET endpoint for current user and provided question id
- [x] Add endpoint to remove current user from alert
- [x] Add tests to ensure users can't edit admin questions
- [x] Add support for deleting alerts
- [x] Add tests for current slack notifications
- [x] Add slack alerts
- [x] Add "first run only" support
- [x] Add progress meter goal support (added but currently broken)
- [x] Verify line chart goal implementation
- [x] Admin should see all alerts for a give question, with the ability to update them
- [x] Non-admin users should only be able to see questions they are a recipient on (currently we also use creator_id)
- [x] Non-admin users can only update questions that they created and are currently an recipient on
- [x] Alert condition has be de-scoped, it should be removed 
- [x] pulse names should be nullable for alerts
- [x] Throw exception when admin calls the unsubscribe route
- [x] Email Scenarios
  - [x] Add an initial "you were subscribed" email
  - [x] Add above/below goal additional text to alert emails
  - [x] Notification of being added to an alert by an admin
  - [x] Notification when a non-admin unsubscribes from an alert
  - [x] Notification when someone unsubscribed you
  - [x]  Notification when an admin deletes an alert you were subscribed to
  - [x] Notification of an alert being deleted due to a change to the question or that question being archived
- [x] Alert deletion scenarios
  - [x] Non-admin unsubscribes from it, and they're the only ones subscribed
  - [x] If a question gets archived directly
  - [x] If a collection gets archived, all containing questions that have alerts
  - [x] After a "first run only" alert is triggered
  - [x] If a rows query is changed to a line/bar/area/progress type
  - [x] When a line/area/bar question has it's goal line removed
  - [x] Switching between line/bar/area is ok, any other change from line/bar/area should be delete
  - [x] If a single breakout changes to multiple breakouts
  - [x] If a question goes from progress bar to anything else
- [ ] Code cleanup to prep for review


Frontend TODOs
-----
- [x] Proper logic around different question and alert types
   - [x] Show the alert menu item in QB only if we have a question with compatible mode and display settings.
   - [x] Restrict the “returns any results” alert type to `SegmentMode` with any visualization
   - [x] Restrict the “crosses a goal line” alert type to `TimeseriesMode` and a line/area/bar visualization with a goal line set.
   - [x] Restrict the “progress bar reaches its goal” alert type to a question with Progress visualization
- [x] Implement alert creation dialog for “crosses a goal line” alert type
- [x] Implement alert creation dialog for “progress bar reaches its goal” alert type (same as the previous from the API perspective; some copy changes in the frontend)
- [x] Apply i18n to UI strings _(some UI strings have formatting / dates so didn't change them yet)_
- [x] For now, persist the "user has seen educational screen about alerts" state using cookies
- [x] In "No Results!" state add "You can also get an alert when there are any results." text
- [x] Show create alert link in alerts list if the current user hasn't created alerts
- [x] fix the double question click issue *fixed*
- [x] if no email/slack not set up, show a dialog. if email set up but no slack set up, then show dialog for non-admins *fixed*
- [x] admins can create alerts even if only pulses enabled *didn't require changes*
- [x] the background color of item should only be blue if there are both non-admin and admin alerts *fixed*
- [x] check the `Uncaught TypeError: Cannot read property 'channel' of undefined` error *fixed*
- [x] don’t let users create an alert if the goal is disabled (the visibility setting is false or something like that) *I think this worked correctly but I still added extra check for seeing that all required visualization settings are present and have expected values*
- [x] Show some kind of "update successful" toast dialog after updating an alert? *added*
- [x] Fix "Edit your alert" title if you are actually editing someone else's alert *it says now just "Edit alert" if it isn't your alert*
- [x] Be able to create/see alerts for native queries too *only raw data alerts were not working; now it lets you create a raw data alert if the native query has a table visualization chosen*
- [x] Prompt the user to save the question if we have a new/dirty question *implemented this properly; hided the alerts button if you are in edit mode of a saved question*
- [x] Add proper error handling to API endpoints and display those in the UI *it was enough to add this to creation/updating forms for now*
- [x] (add loading states where appropriate) *this also applies to creation/updating forms where there is "Loading..." state*
- [x] Write integration tests for different alert types and admin/non-admin workflows
- [ ] Add instrumentation of what kind of alerts are usually created etc
- [x] All sorts of UI polishing

mazameli commented:
Beyond the list of existing to-dos, I noted some QA things that we can address during the polish session tomorrow:

- Do alerts work with existing questions? I.e., if I upgrade to 0.27 in the future, will my existing saved questions show me the option to set up an alert? I ask because when I tried this on Tuesday after spinning up this branch the first time, existing questions didn't seem to be working.

- This is a weird one, but in this branch, clicking on links doesn’t always seem to be working correctly: clicking on saved questions from a list, or a metric from the New Question metric list, or the Custom box on the New Question screen — all of these seem to need to be clicked at least twice to go to their destinations.

- I mentioned this in Slack, but I think we need to tell users that email/slack need to be set up before then can set up an alert. We're doing this with Pulses via modals, and I'm hoping we can reuse those here. For a non-admin to be able to set up an alert, email must be turned on. For an admin to send an alert, technically only Slack needs to be turned on.

- Right now, if a non-admin unsubscribes from an alert they created, the alert is deleted entirely, even if an admin had added additional recipients. We shouldn't delete the alert in this case.

- I’m not 100% sure about this, but it seems that alerts currently stay active even if the question they belong to gets archived. I suspect this because if that question gets _unarchived_, all of the alerts it had on it before it was archived are still there, which I do not think should be the case.

- If I have a goal-line alert on a question, then edit the question and remove the goal line, the alert goes away, which is great. The problem is, if I then later edit the question and put a new goal line on it and save it, my old alert that I thought was deleted gets resurrected. I don’t think this is expected behavior, and we don't notify the user(s) about this possible side effect.

- I confirmed that if a user created alerts, those alerts remain if that user is deleted, which I think is correct. 👍 

- Styling: a non-admin's list of alerts should only make the background of their own alert blue if they're also receiving at least one alert from someone else on the question. I tried to fix this, but it requires a little bit of logic.

- I’m sometimes hitting a bug as an admin when trying to click on “Alerts are on,” and I’m getting this JS error below. I think this happens more regularly, but not always, when there is at least one non-admin alert on a question that an admin also has an alert on. I'll play around with this some more to see if I can figure out how to repro predictably.

```
Uncaught TypeError: Cannot read property 'channel' of undefined
    at AlertListItem.render (AlertListPopoverContent.jsx:136)
    at ReactCompositeComponent.js:793
    at measureLifeCyclePerf (ReactCompositeComponent.js:73)
    at ReactCompositeComponentWrapper._renderValidatedComponentWithoutOwnerOrContext (ReactCompositeComponent.js:792)
    at ReactCompositeComponentWrapper._renderValidatedComponent (ReactCompositeComponent.js:819)
    at ReactCompositeComponentWrapper.performInitialMount (ReactCompositeComponent.js:359)
    at ReactCompositeComponentWrapper.mountComponent (ReactCompositeComponent.js:255)
    at Object.mountComponent (ReactReconciler.js:43)
    at ReactCompositeComponentWrapper.performInitialMount (ReactCompositeComponent.js:368)
    at ReactCompositeComponentWrapper.mountComponent (ReactCompositeComponent.js:255)
```

senior commented:
@mazameli some responses/questions for the polish session

> We still need to implement all the emails

The document you sent out late yesterday has the text and the various emails with scenarios, is that firmed up, or are you still waiting on a 👍 ?

> Right now, if a non-admin unsubscribes from an alert they created, the alert is deleted entirely, even if an admin had added additional recipients. We shouldn't delete the alert in this case.

@attekei is this happening on the UI side, issuing a delete? I would expect that the alert would disappear for that non-admin user based on the rules we've setup, but it shouldn't be deleted.

> I’m not 100% sure about this, but it seems that alerts currently stay active even if the question they belong to gets archived. I suspect this because if that question gets unarchived, all of the alerts it had on it before it was archived are still there, which I do not think should be the case.

This goes along with deleting alerts when they change in a way that no longer makes sense for that alert. Could we come up with a list of updates a question can have that would trigger an alert to be deleted?

> If I have a goal-line alert on a question, then edit the question and remove the goal line, the alert goes away, which is great. The problem is, if I then later edit the question and put a new goal line on it and save it, my old alert that I thought was deleted gets resurrected. I don’t think this is expected behavior, and we don't notify the user(s) about this possible side effect.

My guess is what's happening is that the alert isn't going away (as I haven't implemented this yet, it's still on the TODO list) but maybe the UI is hiding it? Then when the goal is back, it's just not hiding it anymore. The alert is still there and unchanged, just not being shown.

mazameli commented:
Thanks, @senior. Replies:

>>We still need to implement all the emails

>The document you sent out late yesterday has the text and the various emails with scenarios, is that firmed up, or are you still waiting on a 👍 ?

You, @kdoh, and @salsakran all appeared to have reviewed the doc and made comments, so I think it's good to go from my perspective.


> Right now, if a non-admin unsubscribes from an alert they created, the alert is deleted entirely, even if an admin had added additional recipients. We shouldn't delete the alert in this case.

>> @attekei is this happening on the UI side, issuing a delete? I would expect that the alert would disappear for that non-admin user based on the rules we've setup, but it shouldn't be deleted.

This is my expectation as well.

>> I’m not 100% sure about this, but it seems that alerts currently stay active even if the question they belong to gets archived. I suspect this because if that question gets unarchived, all of the alerts it had on it before it was archived are still there, which I do not think should be the case.

>This goes along with deleting alerts when they change in a way that no longer makes sense for that alert. Could we come up with a list of updates a question can have that would trigger an alert to be deleted?

Yes, let me do that right now.

>> If I have a goal-line alert on a question, then edit the question and remove the goal line, the alert goes away, which is great. The problem is, if I then later edit the question and put a new goal line on it and save it, my old alert that I thought was deleted gets resurrected. I don’t think this is expected behavior, and we don't notify the user(s) about this possible side effect.

> My guess is what's happening is that the alert isn't going away (as I haven't implemented this yet, it's still on the TODO list) but maybe the UI is hiding it? Then when the goal is back, it's just not hiding it anymore. The alert is still there and unchanged, just not being shown.

Okay, gotcha.

mazameli commented:
One thing I wanted to explicitly call out since I'm not seeing a to-do item for this is that alerts should be set-up-able on SQL/native queries. The menu UI doesn't currently does not appear on saved SQL questions. Just want to make 100% sure that's on our radar.

![screen shot 2017-11-02 at 2 58 46 pm](https://user-images.githubusercontent.com/2223916/32352417-64183c7e-bfde-11e7-92a5-773e43bf5ed3.png)

kdoh commented on frontend/src/metabase/components/EntityMenu.jsx:
```diff
@@ -25,17 +25,29 @@ class EntityMenu extends Component {
-        open: false
+        open: false,
+        freezeMenu: false,
+        menuItemContent: null
```
Might be nice to add a comment here about what `freezeMenu` and `menuItemContent` are used for. For component state that's not as simple as "open" I've found that useful.

kdoh commented on frontend/src/metabase/components/EntityMenuTrigger.jsx:
```diff
@@ -2,7 +2,7 @@ import React from 'react'
 import cxs from 'cxs'
 
-const EntityzMenuTrigger = ({ icon, onClick, open }) => {
+const EntityMenuTrigger = ({ icon, onClick, open }) => {
```
Hah. Thanks for fixing my Entityz misspelling. 👍

kdoh commented on frontend/src/metabase/query_builder/actions.js:
```diff
@@ -1162,10 +1153,14 @@ export const archiveQuestion = createThunkAction(ARCHIVE_QUESTION, (questionId,
+        dispatch(addUndo(createUndo({
+            type,
+            // eslint-disable-next-line react/display-name
+            message: () => <div> { "Question  was " + type + "."} </div>,
```
Any reason you're not using a template literal here? `Question was ${type}.`

camsaul replied:
where's the i18n!

kdoh commented on frontend/src/metabase/query_builder/components/AlertListPopoverContent.jsx:
```diff
@@ -0,0 +1,254 @@
+const customSpacing = cxs ({
+    marginLeft: '10px'
+})
+const ownAlertCustomSpacing = cxs ({
```
I don't know that we need to specifically call this out as `customSpacing` in the name but It's not a huge deal to me either way.

kdoh commented on frontend/src/metabase/query_builder/components/AlertListPopoverContent.jsx:
```diff
@@ -0,0 +1,254 @@
+        const hasOwnAndOthers = ownAlerts.length > 0 && othersAlerts.length > 0
+
+        return (
+            <div style={{ minWidth: 410 }}>
```
Any reason we're using `style` instead of `cxs` here?

mazameli replied:
I think this was already in there before I gained true cxs enlightenment. I'll change this.

kdoh commented on frontend/src/metabase/query_builder/components/QueryHeader.jsx:
```diff
@@ -431,6 +439,28 @@ export default class QueryHeader extends Component {
+            const updateAlertItem = {
+                title: t`Alerts are on`,
+                icon: "alert",
+                content: (toggleMenu, setMenuFreeze) => <AlertListPopoverContent onCancel={toggleMenu} setMenuFreeze={setMenuFreeze} />
```
Might be worth formatting this a little more nicely to make it a bit easier to read.

kdoh commented on src/metabase/email/alert_you_were_added.mustache:
```diff
@@ -0,0 +1,7 @@
+{{> metabase/email/_header }}
+  <div style="padding: 2em 4em; border: 1px solid #dddddd; border-radius: 2px; background-color: white; box-shadow: 0 1px 2px rgba(0, 0, 0, .08);">
```
It seems silly we have to do the "card" styles for all of these templates. We should consider moving that to a shared partial.

mazameli replied:
True!

camsaul replied:
yeah 👍  there is a lot of repetition going on between the different templates, it would be nice to consolidate them into a partial if possible

kdoh commented on frontend/src/metabase/components/EntityMenu.jsx:
```diff
@@ -25,17 +25,29 @@ class EntityMenu extends Component {
     render () {
         const { items, triggerIcon }  = this.props
-        const { open } = this.state
+        const { open, menuItemContent } = this.state
```
It'd be nice to update `EntityMenu.info.js` with examples of the ability to freeze and have new content to keep things documented.

mazameli commented:
- Confirmed that alerts can be set up on questions that were created and saved before "upgrading" to the alerts branch.
- Found that alerts cannot currently be set up on non-raw-data questions that aren't a time-series-with-goal or a progress bar. One of the conditions in the product doc says alerts should be allowed when, "A card with a non-time-series visualization has 1 or more rows in its result."

mazameli commented:
When using a non-admin account, I'm getting a permissions error when attempting to unsubscribe from an alert that I've created:

![permissions-error](https://user-images.githubusercontent.com/2223916/32576643-563ae3ce-c48c-11e7-848b-87c6f79eb1b0.gif)

mazameli commented:
The email generated when an alert is stopped due to its question being archived has two things funny going on:
1. It's not correctly displaying the name of the now-archived saved question.
2. The link itself goes to `/question/null` instead of e.g. `/question/3`. Of course, that latter URL would simply show a message that the question has been archived. That is what I wrote in the spec, but now I'm thinking it's probably silly to have a link to an archived question. How about we make the title of the question plain text instead of a link after all?

![screen shot 2017-11-08 at 1 44 35 pm](https://user-images.githubusercontent.com/2223916/32576823-e3c7fc0e-c48c-11e7-93ae-fa00498559f9.png)

mazameli commented:
Alerts themselves don't appear to be rendering properly:

![screen shot 2017-11-08 at 3 04 23 pm](https://user-images.githubusercontent.com/2223916/32579519-24f12dc8-c496-11e7-8542-e523c9a61377.png)

mazameli commented:
As an admin, when I add a non-admin as a recipient to an alert, they correctly get an email notification. If I add another admin user, however, they are not receiving an email notification.

mazameli commented:
@attekei After I create/edit/delete an alert, when I'm taken back to the question page, the alerts menu shouldn't still be open:

![screen shot 2017-11-08 at 3 37 11 pm](https://user-images.githubusercontent.com/2223916/32580612-f4262806-c49a-11e7-8004-ea26c40782f1.png)

It can currently lead to weird states like this one:

<img width="757" alt="screen shot 2017-11-08 at 3 38 58 pm" src="https://user-images.githubusercontent.com/2223916/32580621-fd675cc8-c49a-11e7-8132-b4fe0fcf19f8.png">

attekei commented:
Changes I did today:
- Fixed https://github.com/metabase/metabase/pull/6278#issuecomment-342973240 and added a loading status to unsubscription link because it can currently take a while due to email stuff
- Fixed https://github.com/metabase/metabase/pull/6278#issuecomment-342998515. I don't currently close the menu if you exit the creation/updating dialog without actually saving the modal; I can still change this if needed.
- Finished some integration tests I started writing yesterday
- Fixed frontend tests & linter errors

senior commented:
- Fixed https://github.com/metabase/metabase/pull/6278#issuecomment-342991415 it was an issue with me not calling the hiccup render function.
- Fixed https://github.com/metabase/metabase/pull/6278#issuecomment-342974500, it was caused by the a tag in the template not being closed properly, that tag goes away with the switch from link to just text, so it was fixed by making the switch.

mazameli commented:
When switching from a branch other than `alerts` back to the `alerts` branch, I'm sometimes seeing this strange behavior where existing saved questions (possibly only raw data questions which have a filter on them?) will briefly show the menu icon, but then it'll disappear:

![disappearing-menu](https://user-images.githubusercontent.com/2223916/32625944-a57e3a38-c542-11e7-8dd2-8534280df59b.gif)

mazameli commented:
@senior I'm seeing a similar problem with actual alerts as the one [I reported above](https://github.com/metabase/metabase/pull/6278#issuecomment-342974500) where the name of the saved question isn't getting rendered:

![screen shot 2017-11-09 at 12 44 35 pm](https://user-images.githubusercontent.com/2223916/32628823-f57a6896-c54b-11e7-9aad-1075148de763.png)

The happy news is that alerts are rendering. 😎

attekei commented:
I think I was able to fix https://github.com/metabase/metabase/pull/6278#issuecomment-343268632. Other stuff I did today:

- Simplified UI logic related unsubscribing which I realized that I had made overcomplicated
- Added tests that cover alerts list and updating alerts. I think that the most important stuff is covered now
- Found a bug in update API endpoint (unable to change `alert_above_goal` or `alert_first_only` with `PUT /api/alert/:id`) that @senior is fixing
- In no results state there is a `You can also [get an alert] when there are any results.` text for questions that use 'rows present' alert. I had to hide that for new / dirty questions because implementing the "save question first" logic that we have for the QB header alert button would take some time (it would require extracting parts of the `QueryHeader` component and "reduxifying" things that are currently in the local state of QueryHeader).

senior commented:
Pushed up two commits:
- Fixed up the alert_above_goal/alert_first_only bug @attekei found https://github.com/metabase/metabase/pull/6278#issuecomment-343508353
- Fixed up the bug @mazameli found https://github.com/metabase/metabase/pull/6278#issuecomment-343508353

senior commented on resources/migrations/000_migrations.yaml:
```diff
@@ -3939,4 +3939,52 @@ databaseChangeLog:
+                  constraints:
+                    nullable: true
+  - changeSet:
+      id: 71
```
@attekei bumped this from 70. changeset 69 above wasn't released, I could collapse this with the one above and it goes away entirely. People working on this branch thought would need to do some manual `databasechangelog` work though

camsaul replied:
👍  on collapsing these into a single migration

Everyone who's already loaded the branch should probably be able to just do `lein run migrate force` and that should fix things automatically

senior commented on test/metabase/pulse_test.clj:
```diff
@@ -1,16 +1,22 @@
 (ns metabase.pulse-test
-  (:require [expectations :refer :all]
+  (:require [clojure.walk :as walk]
```
I first started working here and I think that test failures here will be more difficult to diagnose than in other tests that make use of the utilities I added to the fake-inbox stuff. I'll still need something like below for slack messages, but I'm having some second thoughts about mocking the send here instead of using the fake-inbox (maybe revisiting [this comment](https://github.com/metabase/metabase/pull/6268#pullrequestreview-71966517))

senior commented:
@attekei There's a lint error on the JS side. I fixed that and found it was hiding a test failure

```
Summary of all failing tests
 FAIL  frontend/test/alert/alert.integ.spec.js (14.788s)
  ● Alerts › alert list for a question › as an admin › should let you edit an alert

    expect(received).toBe(expected)

    Expected value to be (using ===):
      false
    Received:
      true

      at _callee20$ (frontend/test/alert/alert.integ.spec.js:383:64)
      at tryCatch (node_modules/regenerator-runtime/runtime.js:65:40)
      at Generator.invoke [as _invoke] (node_modules/regenerator-runtime/runtime.js:303:22)
      at Generator.prototype.(anonymous function) [as next] (node_modules/regenerator-runtime/runtime.js:117:21)
      at step (frontend/test/alert/alert.integ.spec.js:49:2587)
      at frontend/test/alert/alert.integ.spec.js:49:2747
          at <anonymous>
      at process._tickCallback (internal/process/next_tick.js:188:7)


Test Suites: 1 failed, 12 skipped, 31 passed, 32 of 44 total
Tests:       1 failed, 45 skipped, 241 passed, 287 total
```

I thought the issue might be on the backend so I setup a similar test case. I created an alert with `alert_above_goal` set to true, then issued a PUT to update it to false. It seems to be working for me. Is there something that needs to be changed in the test?

camsaul commented on resources/migrations/000_migrations.yaml:
```diff
@@ -3939,4 +3939,52 @@ databaseChangeLog:
+                  type: varchar(254)
+                  remarks: 'Condition (i.e. "rows" or "goal") used as a guard for alerts'
+                  constraints:
+                    nullable: true
```
things in Liquibase are nullable by default just like when writing DDL statements in SQL. So you don't need to put  `nullable: true` in there

camsaul commented on src/metabase/api/alert.clj:
```diff
@@ -0,0 +1,182 @@
+  (for [alert (if api/*is-superuser?*
+                (pulse/retrieve-alerts-for-card id)
+                (pulse/retrieve-user-alerts-for-card id api/*current-user-id*))
+        :let  [can-read?  (mi/can-read? alert)
```
it doesn't look like you're using the value of `can-read?` here

senior replied:
Good catch, it was a mistake and should be the same as the GET "/" endpoint. I'll refactor and have them use the same function.

camsaul commented on src/metabase/api/alert.clj:
```diff
@@ -0,0 +1,182 @@
+    (assoc alert :read_only (not can-write?))))
+
+(api/defendpoint GET "/question/:id"
+  [id]
```
It's fairly clear what this does but I think the defendpoint macro is still set up to log a warning message if you don't add a docstring to an endpoint. It doesn't trigger the linter because then macro also appends it's own documentation anyway 

Btw when we're adding a bunch of API endpoints it's not a bad idea to run `lein run api-documentation` to update the generated dox. We should probably get that set up to run automatically at some point because I'm always forgetting

senior replied:
Didn't know about `lein run api-documentation`, just ran it and it had a bunch of changes, I'll include it in this PR 👍

camsaul commented on src/metabase/email/messages.clj:
```diff
@@ -226,3 +226,121 @@
+(def ^:private admin-unsubscribed-template (template-path "alert_admin_unsubscribed_you"))
+(def ^:private added-template (template-path "alert_you_were_added"))
+(def ^:private stopped-template (template-path "alert_stopped_working"))
+(def ^:private deleted-template (template-path "alert_was_deleted"))
```
personal preference but I think indenting these all like

```clojure
(def ^:private new-alert-template          (template-path "alert_new_confirmation"))
(def ^:private you-unsubscribed-template   (template-path "alert_unsubscribed"))
(def ^:private admin-unsubscribed-template (template-path "alert_admin_unsubscribed_you"))
(def ^:private added-template              (template-path "alert_you_were_added"))
(def ^:private stopped-template            (template-path "alert_stopped_working"))
(def ^:private deleted-template            (template-path "alert_was_deleted"))
```

looks visually more appealing

camsaul commented on src/metabase/email/messages.clj:
```diff
@@ -226,3 +226,121 @@
+(def ^:private alert-condition-text
+  {:meets "when this question meets its goal"
+   :below "when this question goes below its goal"
+   :rows  "whenever this question has any results"})
```
do these need to be i18n-tagged?

camsaul commented on src/metabase/email/messages.clj:
```diff
@@ -226,3 +226,121 @@
+(defn- alert-results-condition-text [goal-value]
+  {:meets (format "reached its goal of %s" goal-value)
+   :below (format "gone below its goal of %s" goal-value)
+   :rows  "results for you to see"})
```
same thing, is this missing i18n?

camsaul commented on src/metabase/email/messages.clj:
```diff
@@ -226,3 +226,121 @@
+             :sectionStyle render/section-style
+             :colorGrey4   render/color-gray-4
+             :logoFooter   true}
+            (random-quote-context)
```
are we including the random quotes about data in the alert emails?

senior replied:
Yes, all of the emails use [the same footer](https://github.com/metabase/metabase/blob/master/src/metabase/email/_footer.mustache)

camsaul commented on src/metabase/api/pulse.clj:
```diff
@@ -33,9 +33,11 @@
+  [cards]
+  (doseq [card cards
+          :let [card-id (u/get-id card)]]
     (assert (integer? card-id))
```
`u/get-id` already asserts that the output is a valid Integer ID so you don't need to do another `assert` below it

That's sort of the idea behind `u/get-id`: it's guaranteed to return a valid ID. Oh yeah that and it can accept either a map with an `:id` key or an integer which lets you write function that take args like `object-or-id` and handle either case

camsaul replied:
At any rate you should just be able to do `(api/read-check card)` here instead of `(api/read-check Card (u/get-id card))`. 

The latter essentially just translates to `(api/read-check (db/select-one Card :id <id>))` so this will end up doing a DB fetch for each Card. Since you already have the cards those fetches can and should be skipped

camsaul commented on src/metabase/api/card.clj:
```diff
@@ -348,22 +414,40 @@
+          ;; The change doesn't invalidate the alert, do nothing
+          :else
+          nil))
+
```
I think this endpoint has too much going on now. We should move this entire chunk of logic into a separate function and keep the endpoint itself as high-level as possible. I'd prefer a single call to something `(delete-alerts-if-needed! card)` in the endpoint and that function itself could fetch the alerts and do all of the checks there.

camsaul commented:
Omg this PR is crazy big. Wow!

camsaul commented on frontend/src/metabase/alert/alert.js:
```diff
@@ -0,0 +1,147 @@
+}
+
+export const DELETE_ALERT = 'metabase/alerts/DELETE_ALERT'
+const deleteAlertRequest = new RestfulRequest({
```
what is a RestfulRequest? 😱 

I thought we were supposed to be doing like thunk actions and promises and w/e

camsaul commented on frontend/src/metabase/lib/cookies.js:
```diff
@@ -31,6 +32,29 @@ var MetabaseCookies = {
+    setHasSeenAlertSplash: (hasSeen) => {
+        const options = {
+            path: window.MetabaseRoot || '/',
+            expires: 365,
```
This is more of a complaint with `js-cookie` than with your code but it's very confusing to see a cookie being set and then seeing the `expires` field not be a date. My first instinct was that this cookie was being set with bad values or "cooked incorrectly" if you will. IRL `Expires` is supposed to be a date https://developer.mozilla.org/en-US/docs/Web/HTTP/Cookies. 

So I guess maybe it's nice this library is trying to be magical but I just find it confusing. Any chance we can get a comment saying "yes, this is correct" so I don't flip out next time I run into this?

camsaul commented on src/metabase/api/alert.clj:
```diff
@@ -0,0 +1,182 @@
+  ;; Admins are not allowed to unsubscribe from alerts, they should edit the alert
+  (api/check (not api/*is-superuser?*)
+    [400 "Admin user are not allowed to unsubscribe from alerts"])
+  (assert (integer? id))
```
defendpoint already adds regexes for `:id` forcing them to be integer strings and then takes care of parsing them for you so there's no way you could get here and have `id` not be an integer. At any rate if you want to validate params in an endpoint you should use the `{param schema}` map because those get picked up and added to the auto-generated documentation and the produce errors in a format understandable by the frontend, unlike `assert`

camsaul commented on src/metabase/api/collection.clj:
```diff
@@ -45,14 +48,23 @@
+                     archived))
+    (when (and (not (:archived collection-before-update))
+               archived)
+      (let [alerts (apply pulse/retrieve-alerts-for-card (db/select-ids Card, :collection_id id))]
```
what about doing `(when-let [alerts (seq (apply ...))] ...)`
save yourself a LOC

camsaul commented on src/metabase/email/messages.clj:
```diff
@@ -226,3 +226,121 @@
+    (if (true? alert_above_goal)
+      :meets
+      :below)
+    :rows))
```
Nitpicky but I think nested if statements get a little tricky to read in Clojure. One `cond` I think is easier to read:

``clojure
(defn pulse->alert-condition-kwd
  "Given an `ALERT` return a keyword representing what kind of goal needs to be met."
  [{:keys [alert_above_goal alert_condition card creator] :as alert}]
  (cond
    (= "goal" alert_condition) :rows
    (true? alert_above_goal)   :meets
    :else                      :below))
```

Also what's the reason behind checking if `alert_above_goal` is `true`? Is it not always boolean? If it's set to some string or numeric value do we want to treat it as false? That should be noted somewhere IMO otherwise it just looks like the code was written wrong accidentally

camsaul commented on src/metabase/email/messages.clj:
```diff
@@ -226,3 +226,121 @@
+        message-ctx  (default-alert-context alert (alert-results-condition-text goal-value))]
+    (render-message-body "metabase/email/alert"
+                         (assoc message-ctx :pulse body :firstRunOnly? alert_first_only)
+                         (seq @images))))
```
it looks like about half of this function (the `images` and `body` part) is just copy-pasta from `render-pulse-email`. Maybe there's some way we could avoid repeating the logic between these two `render-` functions and have them both call a shared function that takes care of this stuff?

camsaul commented on src/metabase/email/messages.clj:
```diff
@@ -226,3 +226,121 @@
+(defn send-you-unsubscribed-alert-email!
+  "Send an email to `WHO-UNSUBSCRIBED` letting them know they've unsubscribed themselves from `ALERT`"
+  [alert who-unsubscribed]
+  (send-email! who-unsubscribed "You unsubscribed from an alert" you-unsubscribed-template
```
i18n?

camsaul commented on src/metabase/email/messages.clj:
```diff
@@ -226,3 +226,121 @@
+  (let [subject (format "%s %s added you to an alert" first_name last_name)]
+    (send-email! user-added subject added-template (default-alert-context alert alert-condition-text))))
+
+(def ^:private not-working-subject "One of your alerts has stopped working")
```
I think basically this whole namespace needs to be i18n'ed

senior replied:
I didn't i18n this namespace intentionally. These strings are either email subjects or are snippets of text that are being injected into the emails. We don't currently have an i18n setup for these email templates, that I'm aware of. I was thinking if we i18n this, and had a translation, we would be in a situation where we have a Spanish email subject with a mostly English email body, but with a few Spanish phrases like 'results for you to see' thrown in. That seemed worse that an all English email.

I was thinking we'd need to address i18n for emails as a whole before i18n'ing this.

camsaul commented on src/metabase/models/pulse.clj:
```diff
@@ -85,22 +87,96 @@
+                           [:= :pcr.user_id user-id]]})))
+
+(defn retrieve-alerts-for-card
+  "Find all alerts for `CARD-ID`, used for admin users"
```
`card-id` isn't a param for this function 😕

camsaul commented on src/metabase/models/pulse.clj:
```diff
@@ -85,22 +87,96 @@
+                  :join   [[PulseCard :pc] [:= :p.id :pc.pulse_id]]
+                  :where  [:and
+                           [:not= :p.alert_condition nil]
+                           [:in :pc.card_id card-ids]]})))
```
If you're using `:in` you have to make sure you only do the query in the first place if the seq is non-empty. HoneySQL is dumb and will generate a clause like `WHERE pc.card_id IN ()` which will barf. But you don't need to do the query anyway because nothing is going to come back. So make sure you wrap this whole thing in a `(when (seq card-ids) ...)`

camsaul commented on src/metabase/pulse.clj:
```diff
@@ -44,24 +51,13 @@
+(def ^:private alert-notification-condition-text
+  {:meets "reached its goal"
+   :below "gone below its goal"
+   :rows  "results"})
```
i18n

camsaul commented:
Maybe this is a dumb question, but do we have any tests to see if alerts actually get sent? It looks like all the logic behind creating/updating/deleting alerts and related objects is tested solidly but (maybe I missed it) I didn't see anything that:

1. created an alert
2. changed some data somewhere
3. checked if an alert was or wasn't fired off

senior commented:
@camsaul I did create one [here](https://github.com/metabase/metabase/pull/6278/files#diff-d9511ad8d69f13e04625404a15d31813), though unfortunately not until after @mazameli found a bug. It's covering that we can get the alert out of the db when triggered and call the pulse code correctly.

senior commented:
I just pushed up fixes based on the code review feedback,thanks @camsaul !

As part of that, I collapsed the two database migrations and put the correct next changeset id in there. If you've been working off of this branch and attempt to start up again, it will probably fail as the database checksums are different. To fix this, if you're using h2, run `lein h2` or open up the mysql/postgresql console:

Changeset 71 has been removed, you should remove it too:
```
delete from datatabasechangelog where id='71';
```
Then update the checksum for the old version of changeset 69:
```
update databasechangelog set md5sum='7:1de67869247405ada4cd35068f11cb5f' where id='69';
```

BTW, changing the md5sum is a bit of a pain, so I created https://github.com/metabase/metabase/issues/6372 as it seems like it should be pretty easy to automate.

attekei committed:
Alerts Frontend

This commit has the UI for users to add alerts to a question. Admins
are able to setup alerts for questions and modify alerts that users
have setup.

iethree committed:
Alerts Backend

This commit has the backend code needed to support alerts. From the
frontend perspective, Alerts is separate from pulses and has a
different endpoint, however much of the backend code and database
tables are shared. A few new columns were added to the pulse table to
support alerts, but most of the data is the same between pulses and
alerts.

This also adds several "meta-notifications" around alerts to let users
know when they've been added to an alert, removed from an alert or if
a question has changed such that the alert is no longer meaningful.

This fixes #1372, fixes #6261 and fixes #5768.

urielka commented:
Hi, when is this going to be released? looks awesome!

mazameli commented:
@urielka This will be shipped in our upcoming 0.27 release, which will be coming out imminently.

satishvarmadandu commented:
Is there any way to setup alerts every few mins. From the UI, we see minimum frequency is 1 hr to schedule alerts
